esmvaltool/diag_scripts/quantilebias/quantilebias.R

```diff
@@ -0,0 +1,144 @@
+# #############################################################################
+# quantilebias.r
+# Authors:       E. Arnone (ISAC-CNR, Italy)
+# 	         S. Terzago (ISAC-CNR, Italy)
+# 	         J. von Hardenberg (ISAC-CNR, Italy)
+# #############################################################################
+# Description
+# ESMValTool diagnostic for calculation of the precipitation quantile bias
+# following Mehran et al. (2014)
+#
+# Required
+# - #CDO
+# - observational monthly mean global precipitation climatology
+# - (e.g. from the GPCP project: http://gpcp.umd.edu)
+#
+# Optional
+#
+# Caveats
+#
+# Modification history
+#    20180926-A_arno_en: Refined for usage as recipe
+#    20180518-A_arno_en: Written for v2.0
+#
+# #############################################################################
+
+library(tools)
+library(yaml)
+
+# read settings and metadata files
+args <- commandArgs(trailingOnly = TRUE)
+settings <- yaml::read_yaml(args[1])
+metadata <- yaml::read_yaml(settings$input_files)
+for (myname in names(settings)) {
+  temp <- get(myname, settings)
+  assign(myname, temp)
+}
+
+# get name of climofile and list associated to first climofile
+climofiles <- names(metadata)
+climolist <- get(climofiles[1], metadata)
+
+# say hi
+diag_base <- climolist$diagnostic
+print(paste0(diag_base, ": starting routine"))
+
+# get variable name
+varname <- climolist$short_name
+
+# create working dirs if they do not exist
+dir.create(work_dir, recursive = T, showWarnings = F)
+setwd(work_dir)
+
+# setup provenance file and list
+provenance_file <- paste0(run_dir, "/", "diagnostic_provenance.yml")
+provenance <- list()
+
+# extract metadata
+models_name <- unname(sapply(metadata, "[[", "dataset"))
+reference_model <- unname(sapply(metadata, "[[", "reference_dataset"))[1]
+models_start_year <- unname(sapply(metadata, "[[", "start_year"))
+models_end_year <- unname(sapply(metadata, "[[", "end_year"))
+models_experiment <- unname(sapply(metadata, "[[", "exp"))
+models_ensemble <- unname(sapply(metadata, "[[", "ensemble"))
+
+ref_idx <- which(models_name == reference_model)
+ref_data_file <- climofiles[ref_idx]
+
+## Loop through input models
+for (model_idx in c(1:(length(models_name)))) {
+  if (model_idx == ref_idx) {
+     next
+  }
+  # Setup parameters and path
+  exp <- models_name[model_idx]
+  year1 <- models_start_year[model_idx]
+  year2 <- models_end_year[model_idx]
+  infile <- climofiles[model_idx]
+  model_exp <- models_experiment[model_idx]
+  model_ens <- models_ensemble[model_idx]
+
+  inregname <- paste0(exp, "_", model_exp, "_", model_ens, "_",
+                      toString(year1), "-", toString(year2), "_", varname)
+  outfile <- paste0(work_dir, "/", inregname, "_", perc_lev, "qb.nc")
+  print(paste0(diag_base, ": pre-processing file: ", infile))
+
+  print(paste0(diag_base, ": ", perc_lev, " percent quantile"))
+  system(paste0("cdo selvar,", varname, " ", infile, " tmp_model.nc"))
```

Comment:
This does not check what the return value of the command is. If a problem occurs, the script will keep running and finish successfully. Could you address this? Maybe by writing a single function that you can pass cdo arguments instead of many `system("cdo ...")` calls?

From the [`system` manual](https://stat.ethz.ch/R-manual/R-devel/library/base/html/system.html), it also appears that it's better to use the `sytem2` command instead:
> This interface has become rather complicated over the years: see system2 for a more portable and flexible interface which is recommended for new code.

Reply:
Ok, I wrote a small interface similar to python-cdo, using system2, with some minimal error checking. I also improved the treatment of temporary files. What do you think @bouweandela ?

Reply:
Thanks, it looks much better now. Maybe you could automatically use `tempfile()` in the function if no `output` argument is specified, but it's fine like this.

Reply:
Actually there are cdo commands which require no output file, so the interface is more general and suitable for all possible commands as it is now. I will probably make a better version in the future (for example returning the text output of commands) but all this is not needed here.